store/src/db/interface.rs

```diff
@@ -0,0 +1,7 @@
+use crate::types::BlockNumber;
+use miden_crypto::hash::rpo::RpoDigest;
+
+#[tonic::async_trait]
+pub trait Db {
+    async fn get_nullifiers(&mut self) -> Result<Vec<(RpoDigest, BlockNumber)>, anyhow::Error>;
```

Comment:
Is there a way to return an iterator rather than a vector? This way, we won't have to materialize the vector in memory.

Another option might be to change this to something like `build_nullifier_tree(&mut self) -> TieredSmt` - but I'm not sure if this is a good idea.

Reply:
> Is there a way to return an iterator rather than a vector?

It should be doable like this:

- create a channel
- spawn a new async thread to kick off the query, and pass the producer end of the channel
   - the channel will be moved to the closure, which will execute the query on the thread pool and stream data in
- then on the original async thread, consume from the channel

I'd need to look up channel implementations, to find one that can have a producer in a blocking context, and a consumer on an async context. And would probably need to change the TieredSmt to consume from an async iterator.

> Another option might be to change this to something like build_nullifier_tree(&mut self) -> TieredSmt

This seems simpler to me

Reply:
tokio mpsc does support mixing sync/async code:

On the sender: [fn blocking_send](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.blocking_send) 
On the receiver: [async fn recv](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.recv)

If we decide to go on that route, it probably makes sense to drop the use of the connection pool library and just spawn the SQLite thread ourselves.

Reply:
Thanks! Yes, not something we need to do now - but good to know that we have an option for the future.